### PR TITLE
bpo-34454: Clean up datetime.fromisoformat surrogate handling

### PR DESCRIPTION
**Pull request by pganssle**
This is a fixup PR for #8862, per @serhiy-storchaka's comments. I have addressed two PEP-7 violations, switched to using `_PyUnicode_Copy` in `_sanitize_fromisoformat_str`, and fixed an issue where the sanitized string (rather than the original string) was displayed as part of the error message.

Additionally, I noticed that the pure Python implementation uses the equivalent of `%U` instead of `%R` in its error printing, so I switched all the `fromisoformat` errors over to using `%U` for consistency.

[bpo-34454](https://www.bugs.python.org/issue34454)

https://bugs.python.org/issue34454

### CODE REVIEW

**Comment by pganssle** [comment]
CC @taleinat

File: Lib/test/datetimetester.py
Comment (taleinat) [inline Lib/test/datetimetester.py]: IMO just checking that `ValueError` is raised is enough.

**Review by taleinat — APPROVED** [review]
LGTM, one small comment.

File: Lib/test/datetimetester.py
Comment (pganssle) [reply Lib/test/datetimetester.py]: The regex is actually the point of the test. Before these changes, the error message was accidentally including the *sanitized* string, so `2018-01-03\ud80001:0113` would give you something like `Invalid isoformat str: 2018-01-03T01:0113`.

As a result of this test, I also realized that the C and pure Python versions had slightly different error messages, so I corrected that (and this test enforces it).

If we don't care to enforce any particular conditions on the error message, we can drop the whole test, since the part about raising `ValueError` has pretty good test coverage elsewhere.

**Review by zhangyangyu — COMMENTED** [review]
Is the `needs_decref` flag a must to accomplish the logic? Isn't `Py_XDECREF` enough?

**Comment by serhiy-storchaka** [comment]
`PyUnicode_AsUTF8AndSize()` can fail not only due to a lone surrogate, but because of MemoryError. In this case the exception should not be replaced with a ValueError.

There is similar issue in the pure Python version. Don't use `except Exception`, it can hide unexpected errors.

I suggested to use `%R` in the error message because including the raw string can be confusing in the case of empty string, or string containing trailing whitespaces, invisible or unprintable characters. It is better to change the pure Python version.

While we are here, please make other fromisoformat related code PEP-7 compliant: move top-level `{` on separate line, break a line after `}` in `} else`, break a line after `static int` in `static int parse_isoformat_date`, add spaces around operators (e.g. `rv?-5:1` => `rv ? -5 : 1`).

Also, the parsing code can be simplified. Since different error codes are not distinguished, it is enough to use just `-1` and merge conditions.

**Comment by serhiy-storchaka** [comment]
I would use the following code instead of `_sanitize_isoformat_str`:
```c
    PyObject *bytes = NULL;
    Py_ssize_t len;
    const char * dt_ptr = PyUnicode_AsUTF8AndSize(dtstr, &len);

    if (dt_ptr == NULL) {
        bytes = _PyUnicode_AsUTF8String(dtstr, "surrogatepass")
        if (bytes == NULL) {
            return NULL;
        }
        dt_ptr = PyBytes_AS_STRING(bytes);
        len = PyBytes_GET_SIZE(bytes);
    }

    ...

    Py_XDECREF(bytes);
```

**Comment by pganssle** [comment]
> Is the needs_decref flag a must to accomplish the logic? Isn't Py_XDECREF enough?

@zhangyangyu Indeed no, because `_sanitize_isoformat_str` returns the original string in *most* cases, and the reference to `dtstr` is owned by the calling function.

> I would use the following code instead of `_sanitize_isoformat_str`:

@serhiy-storchaka That is similar to the code I was using originally, but this version is much faster in certain cases. I think you can see the reasoning in the original PR, #8862.

> `PyUnicode_AsUTF8AndSize()` can fail not only due to a lone surrogate, but because of `MemoryError`. In this case the exception should not be replaced with a `ValueError`.

Good point.

> I suggested to use %R in the error message because including the raw string can be confusing in the case of empty string, or string containing trailing whitespaces, invisible or unprintable characters. It is better to change the pure Python version.

Fine by me, I was mainly worried about the consistency.

> While we are here, please make other fromisoformat related code PEP-7 compliant: move top-level { on separate line, break a line after } in } else, break a line after static int in static int parse_isoformat_date, add spaces around operators (e.g. rv?-5:1 => rv ? -5 : 1).

Thanks for the tips. I don't do enough PEP 7 C programming to have these rules internalized - is there a standard code formatter (or even a linter) I can use that you recommend?

> Also, the parsing code can be simplified. Since different error codes are not distinguished, it is enough to use just -1 and merge conditions.

The reason for the different error codes in the parsing is that in an alternate version of this code (a full-spec ISO8601 parser), I gave more detailed error messages, e.g. "error parsing time" and "error parsing time zone", etc, which required more insight into where the error happened. I decided that for the moment the error should *not* be that specific to minimize the API maintenance burden, but because we might want to give a richer error message in the future and the work is already done to support it, I decided to leave it. I don't think it hurts anything to leave the return values as is.

File: Lib/test/datetimetester.py
Comment (serhiy-storchaka) [inline Lib/test/datetimetester.py]: Or just `str(cm.exception)`.

**Comment by pganssle** [comment]
I used `clang-format` with [this PEP 7-style settings file](https://github.com/closeio/ciso8601/blob/master/.clang-format) (coincidentally from an iso 8601 parser, the first google result for a PEP 7 clang-format file). I don't think there's any guarantee that I got all the PEP 7 violations (plus `clang-format` changed a *lot* more than my code in that file, so I had to filter it out manually). @serhiy-storchaka @abalkin Please let me know if you spot any more violations.

File: Lib/test/datetimetester.py
Comment (pganssle) [inline Lib/test/datetimetester.py]: I was having trouble using a `assertRaisesRegex` here, I think because of the escape characters, but `in` works fine.

**Comment by zhangyangyu** [comment]
> Indeed no, because _sanitize_isoformat_str returns the original string in most cases, and the reference to dtstr is owned by the calling function.

Sorry I don't get you. I mean something like:

```
PyObject *dtstr_clean = _sanitize_isoformat_str(dtstr);
if (duster_clean == NULL) goto error;  // or just `return NULL` and `Py_DECREF` in error
...
Py_DECREF(tz_info);
Py_DECREF(dtstr_clean);
...
error:
    Py_XDECREF(dtstr_clean);
```

I think it's easier to understand and maintain.

**Comment by pganssle** [comment]
@zhangyangyu `_sanitize_fromisoformat_str(dtstr)` when `dtstr = "2018-01-01"` will return `dtstr`, not `NULL`. Since we have a borrowed reference to `dtstr` we cannot decrement the reference count.  `dtstr_clean` is `NULL` only if an error has occurred in `_sanitize_fromisoformat_str`. In most cases it returns the input (since most inputs are pre-sanitized), but occasionally it returns a new object, and `datetime_fromisoformat` *must* own that object. `needs_decref` is a flag to indicate whether or not `datetime_fromisoformat` owns `dtstr_clean`.

Consider if I make the following change:

```diff
diff --git a/Modules/_datetimemodule.c b/Modules/_datetimemodule.c
index 68b1b2f501..78c08bdb19 100644
--- a/Modules/_datetimemodule.c
+++ b/Modules/_datetimemodule.c
@@ -4975,18 +4975,14 @@ datetime_fromisoformat(PyObject *cls, PyObject *dtstr)
                                             second, microsecond, tzinfo, cls);
 
     Py_DECREF(tzinfo);
-    if (needs_decref) {
-        Py_DECREF(dtstr_clean);
-    }
+    Py_DECREF(dtstr_clean);
     return dt;
 
 invalid_string_error:
     PyErr_Format(PyExc_ValueError, "Invalid isoformat string: %R", dtstr);
 
 error:
-    if (needs_decref) {
-        Py_DECREF(dtstr_clean);
-    }
+    Py_XDECREF(dtstr_clean);
 
     return NULL;
 }
```

After running `make`, I do:

```
$ ./python -c "from datetime import datetime; print(datetime.fromisoformat('2018-01-01'))"
Fatal Python error: Objects/listobject.c:341 object at 0x7f4dc6ef5f20 has negative ref count -2604246222170760230

Current thread 0x00007f4dc742b080 (most recent call first):
Aborted (core dumped)
$ ./python -c "from datetime import datetime; print(datetime.fromisoformat('2018-01-01\ud80001:01:01'))"
2018-01-01 01:01:01
```

The segfault is because in the first case `datetime_fromisoformat` had a `dtstr_clean` pointing to `dtstr`, a borrowed reference. In the second case, `dtstr_clean` pointed to a temporary object owned by `datetime_fromisoformat`.

**Comment by zhangyangyu** [comment]
@pganssle Sorry I forget mention we need to `Py_INCREF(dtstr)` in `_sanitize`.

**Comment by pganssle** [comment]
@zhangyangyu Ah, yeah, that's an elegant solution.

File: Lib/test/datetimetester.py
Comment (zhangyangyu) [reply Lib/test/datetimetester.py]: You could use `assertRaisesRegex(ValueError, re.escape(f"{dtstr!r}"))`.

Comment (pganssle) [reply Lib/test/datetimetester.py]: Ah, thank you! I thought there must be a way to do this but it seemed like a lot of work to figure it out. Thanks for the pointer. :smile:

**Review by izbyshev — APPROVED** [review]
LGTM!

File: Lib/test/datetimetester.py
Comment (taleinat) [inline Lib/test/datetimetester.py]: Perahps just `re.escape(repr(dtstr))`?

Comment (pganssle) [reply Lib/test/datetimetester.py]: Oops, thanks. This is a legacy from the fact that I wasn't sure if this had to match at the beginning of a line, so it was `f".*{dtstr!}"`.

Comment (pganssle) [reply Lib/test/datetimetester.py]: Fixed. Thanks.

**Comment by pganssle** [comment]
@serhiy-storchaka Ping. Everything up to your standards?

**Review by vstinner — CHANGES_REQUESTED** [review]
Mixing a bugfix and code cleanup is a bad practice: it becomes harder to cherry-pick the fix to older versions, it's harder to revert the change in case of failure, etc.

Please revert all coding style changes and create a new PR once this one is merged.

I'm ok with a PR only fixing coding style, since I like all your coding styles changes, it really enhances the code :-)

**Comment by pganssle** [comment]
> Mixing a bugfix and code cleanup is a bad practice: it becomes harder to cherry-pick the fix to older versions, it's harder to revert the change in case of failure, etc.

This is a problem with squash merges. All my PRs consist of commits with a curated history of incremental atomic changes for just this reason.

The latency between PR proposal and merging is so long, as a general rule, that making small, atomic PRs *depending on one another* is incredibly difficult. Worth considering the incentives here.

**Comment by serhiy-storchaka** [comment]
This code is new in 3.7 so I don't see a problem with a code cleanup.

**Comment by vstinner** [comment]
I am fine with code cleanup. I just ask to do two PRs, so it's easier for me to review it. Right now it's unclear to me which part fixes the issue.

**Comment by pganssle** [comment]
@vstinner [`5d713e05`](https://github.com/python/cpython/pull/8959/commits/5d713e052556a2b75745491c3808191b298a66dd) and [`2bcd8ef4`](https://github.com/python/cpython/pull/8959/commits/2bcd8ef4834d2fa1c0553deb28211c6a64fb2760) are formatting-only changes. The others are substantive changes to how the algorithm works:

- [`d9f95666dc`](https://github.com/python/cpython/pull/8959/commits/d9f95666dc405b51fc4ed294237f1ed84cbf7417) is a performance improvement
- [`5ee8d212`](https://github.com/python/cpython/pull/8959/commits/5ee8d2126c1f4fa557b2b7a27db8114f27742aee)  changes how the error format slightly
- [`c081e152`](https://github.com/python/cpython/pull/8959/commits/c081e1525d3ad0fdb522ef68dcba66f4ab9a6602) fixes a bug in error formatting
- [`eb606515`](https://github.com/python/cpython/pull/8959/commits/eb60651541a49da8433e4767586b119188393d7e) fixes a bug where erroneous errors might be raised if an operation fails for some reason other than bad unicode decode (e.g. `MemoryError`)
- [`b1dd7b36`](https://github.com/python/cpython/pull/8959/commits/b1dd7b36cc8ef33e49ba0609b258f2c285cbe4bd) does the same thing as `5ee8d212` but for the Python implementation.
- [`a19603f9`](https://github.com/python/cpython/pull/8959/commits/a19603f999c674cf76ea97933689c28d7778fc77) refactors the ownership logic on `dtstr` to avoid carrying around ownership in a separate variable.

If you're concerned with having clean atomic commits that can be reverted independently in the final commit history, then I think squashing the style changes into 1 commit, the "error format should be a repr" changes into 1 commit, and leaving all the other commits alone, as they are already atomic. Then you can do a non-squash merge and not bother with the pain of creating two or more separate PRs.

**Comment by vstinner** [comment]
> I have rewritten the branch's history to group together  related changes in atomic commits.

Hum. I forgot to explain that CPython policy doesn't allow to merge a PR, only to squash changes into one unique commit. That's why I'm asking for a second PR.

**Comment by pganssle** [comment]
> Hum. I forgot to explain that CPython policy doesn't allow to merge a PR, only to squash changes into one unique commit. That's why I'm asking for a second PR.

Your reasoning for why you want two PRs applies equally well to each individual atomic commit, not just to "style changes" and "bug fixes" as mentioned in my comments, in fact, it's even *more important* not to squash multiple commits that have actual effects on the program than it is to avoid squashing style changes together with behavior changes. If you squash style changes together and have to revert, you at least aren't changing unrelated things. If you squash two bugfixes together, you can't revert one without reverting the other.

 If you want to prioritize process here, I can make 6 PRs, or I can make 4 PRs, or we can do a squash merge on what exists now and not worry about it too much. That said, I don't see the point of doing such a thing. The point of a squash merge policy is because it keeps a relatively clean history while avoiding a lot of nitpicking about re-writing the history to squash fixups and the like. Given that I **already** have created a clean history, none of the downsides of doing a regular merge exist for this PR, but splitting it up into multiple PRs has all of the downsides of a "regular merge with clean histories" policy.

**Comment by pganssle** [comment]
Feel free to merge or not merge as desired. You're also welcome to cherry pick my commits into whatever number of PRs you want.

**Comment by vstinner** [comment]
I read individual commits and now I'm confused which parts are related to the bugfix or not. I let you (@pganssle) reorganize commits for that :-)

```
   Py_ssize_t len = PyUnicode_GetLength(dtstr);
+    if (len < 0) {
+        return NULL;
+    }
```

This change looks a bugfix, so I suggest to include it in the bugfix PR as well.

**Comment by pganssle** [comment]
You seem to have a pretty good idea of what you want this to look like, so I'm not terribly interested in doing a guess-and-check. IMO the only style-only commit is d809d712c3c79ccadc6061469f9751f7120edbf5. Everything else actually changes the behavior of the program.

If you're going to merge all the bug fixes together, might as well merge the style changes with them and if something needs reversion I'll just make a partial reversion PR or something. Seems like less work than rewriting the history and cherry-picking out separate PRs just on the off chance that one of these things needs to be reverted.

**Comment by taleinat** [comment]
I think we can still "rebase and merge", just not via the GitHub UI. I don't mind doing it myself. No need for additional PRs, IMO. Please stop spending time on this discussion.

**Review by serhiy-storchaka — APPROVED** [review]

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
Note: I agree that this change must be backported to 3.7 (even if it's mostly "cleanup") to ease future bugfixes in fromisoformat().

**Comment by pganssle** [comment]
@vstinner I think the new test actually *does* fix a bug resulting from the PR for which this is a cleanup, but there are a few other bugs fixed in there for which no test is possible.

The bug fixed where adding a test *was* possible was that you could pass an invalid string like this: `"2018-01-03\ud80001:0113"` to `datetime.fromisoformat`, and the error message would display the string with an error in it as `"2018-01-03T01:0113"` instead - it was leaking an implementation detail.

The first commit in the PR also fixes a null dereference bug IIRC. One of the exception-handling commits also fixes a bug with over-zealous error catching.

**Comment by vstinner** [comment]
@taleinat: "I think we can still "rebase and merge", just not via the GitHub UI."

How would you proceed?

**Comment by vstinner** [comment]
I merged this PR. I has been approved by 4 developers including 3 core developers, I agree with @taleinat: it has been discussed enough :-)

**Comment by vstinner** [comment]
Note: I ran " ./python -m test -R 3:3 test_datetime" before merging the PR, and regrtest didn't find any reference leak.

**Comment by taleinat** [comment]
> 
> 
> @taleinat: "I think we can still "rebase and merge", just not via the GitHub UI."
> 
> How would you proceed?

For future reference: Just use a git client e.g. the cli; rebase, merge into master, push master.

(... unless  GitHub's master branch is now protected and we can't push directly into it?)

**Comment by taleinat** [comment]
> > @taleinat: "I think we can still "rebase and merge", just not via the GitHub UI."
> > How would you proceed?
> 
> For future reference: Just use a git client e.g. the cli; rebase, merge into master, push master.
> 
> (... unless GitHub's master branch is now protected and we can't push directly into it?)

So apparently I was wrong. Currently, all of our long-lived branches are protected, and the only way to merge a PR as two commits is to create an extra PR.

**Pull request merged**
